worksheet_D/oxo.py

```diff
@@ -0,0 +1,121 @@
+class OxoBoard:
+    def __init__(self):
+        self.board = {(0,0):0,(0,1):0,(0,2):0,(1,0):0,(2,0):0,(1,1):0,(2,2):0,(1,2):0,(2,1):0}
+    def get_square(self, x, y):
+        """ Return 0, 1 or 2 depending on the contents of the specified square. """
+        return self.board[x, y]
+
+
+    def set_square(self, x, y, mark):
+        """ If the specified square is currently empty (0), fill it with mark and return True.
+            If the square is not empty, leave it as-is and return False. """
+        if self.board[x,y] == 0:
+            self.board[x,y] = mark
+            return True
+        else:
+            return False
+
+
+
+
+    def is_board_full(self):
+        """ If there are still empty squares on the board, return False.
+            If there are no empty squares, return True. """
+        freespaces = 9
+        for i in xrange(0,3):
+            for j in xrange(0,3):
+                if self.board[i, j] > 0:
+                    freespaces -= 1
+                if freespaces == 0:
+                    return True
+        return False
+
+    def get_winner(self):
+        """ If a player has three in a row, return 1 or 2 depending on which player.
+            Otherwise, return 0. """
+        if self.board[1, 1] != 0:
+            if self.board[1, 1] == self.board[0, 0]:
+                if self.board[2, 2] == self.board[1, 1]:
```

Comment:
This could be made neater by combining the conditions with `and` rather than using nested `if`s, or by using Python's ability to chain `==` and `!=` together to compare more than two values at once.